### PR TITLE
PUB-1384 Crown Firm List Pdf Template

### PR DESCRIPTION
**Pull request by junaidiqbalmoj**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/PUB-1384

### Change description ###

Pdf template for Crown Firm list.

image

image

image

image

**Does this PR introduce a breaking change?** (check one with "x")
[ ] Yes
[x] No

### CODE REVIEW

**Review by joshblackmoor — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java [hunk 1]
```diff
@@ -217,55 +216,88 @@ private static void formatCaseHtmlTableCrownDailyList(JsonNode hearing) {
...
-                        formatPartyInformation(party, prosecutingAuthority);
-                        break;
+        StringBuilder defendantRepresentative = new StringBuilder();
+
+        if (hearing.has("party")) {
+            hearing.get("party").forEach(party -> {
+                if (!GeneralHelper.findAndReturnNodeText(party, "partyRole").isBlank()) {
+                    switch (PartyRoleMapper.convertPartyRole(party.get("partyRole").asText())) {
+                        case "PROSECUTING_AUTHORITY": {
+                            formatPartyInformation(party, prosecutingAuthority, true);
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java:229]: Thought we discussed last week and agreed prosecuting authority and dependant representative are coming from organisation details and not individual details? This is going to be in conflict with the PUB-1607 where I changed all the other crime lists to use organisation details for these 2 fields. Happy to discuss this but we do need to keep them consistent as all crime lists look quite similar in those fields

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java [hunk 2]
```diff
@@ -217,55 +216,88 @@ private static void formatCaseHtmlTableCrownDailyList(JsonNode hearing) {
...
+        }
+
+        return (title + " "
+            + surname
+            + separator
+            + forName + " "
+            + middleName).trim();
+    }
+
+    private static String formatNameWitForNameFirst(String title, String forName, String middleName, String surname) {
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java:291]: typo in method name

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java [hunk 3]
```diff
@@ -217,55 +216,88 @@ private static void formatCaseHtmlTableCrownDailyList(JsonNode hearing) {
...
+        String separator = " ";
+        if (!forName.isEmpty() || !middleName.isEmpty()) {
+            separator = ", ";
+        }
+
+        return (title + " "
+            + surname
+            + separator
+            + forName + " "
+            + middleName).trim();
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java:288]: The example in the mock is "surname, forename" only. The frontend lust template is also "surname, forename". Should we include title and middlename?

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    private static List<String> findUniqueSittingDatesPerCounts(JsonNode artefact) {
+        Map<Date, String> allSittingDateTimes = new ConcurrentHashMap<>();
+        artefact.get(COURT_LIST).forEach(courtList -> {
+            Map<Date, String> sittingDateTimes = EtFortnightlyPressListHelper.findAllSittingDates(
+                courtList.get(LocationHelper.COURT_HOUSE).get(COURT_ROOM));
+            allSittingDateTimes.putAll(sittingDateTimes);
+        });
+        return EtFortnightlyPressListHelper.findUniqueDateAndSort(allSittingDateTimes);
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java:53]: If this method is required for both ET list and Crown Firm list. We should move it out to a common class

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        context = preprocessArtefactForThymeLeafConverter(artefact, metadata, language, true);
+        crownFirmListFormatted(artefact);
+        splitByCourtAndDate(artefact);
+        return context;
+    }
+
+    private static List<String> findUniqueSittingDatesPerCounts(JsonNode artefact) {
+        Map<Date, String> allSittingDateTimes = new ConcurrentHashMap<>();
+        artefact.get(COURT_LIST).forEach(courtList -> {
+            Map<Date, String> sittingDateTimes = EtFortnightlyPressListHelper.findAllSittingDates(
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java:49]: If this method is required for both ET list and Crown Firm list. We should move it out to a common class

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+            });
+            courtListByDateArray.add(courtListArray);
+        }
+        ((ObjectNode)artefact).putArray("courtListsByDate").addAll(courtListByDateArray);
+    }
+
+    private static void checkAndAddToArrayNode(ArrayNode arrayToCheck, ObjectNode destinationNode,
+                                               String destinationNodeAttribute, ArrayNode arrayToAdd) {
+        if (arrayToCheck.size() > 0) {
+            (destinationNode).putArray(destinationNodeAttribute).addAll(arrayToCheck);
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java:107]: Why is the brackets required around 'destnationNode'? It is not cast to anything. Same for other similar calls in this class

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    public static void crownFirmListFormatted(JsonNode artefact) {
+        artefact.get(COURT_LIST).forEach(courtList -> {
+            courtList.get(LocationHelper.COURT_HOUSE).get(COURT_ROOM).forEach(courtRoom -> {
+                courtRoom.get("session").forEach(session -> {
+                    session.get(SITTINGS).forEach(sitting -> {
+                        String sittingDate = DateHelper.formatTimeStampToBstHavingWeekDay(
+                            sitting.get(SITTING_START).asText(),
+                            "dd MMMM yyyy", Language.ENGLISH);
+                        ((ObjectNode)sitting).put(SITTING_DATE, sittingDate);
+                        MagistratesStandardListHelper.manipulatedSitting(courtRoom, session, sitting,
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java:140]: Other list helper class is independent to this one. If the helper method is shared, it needs to be moved out to a shared class

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+                                                                         FORMATTED_COURT_ROOM);
+                        sitting.get("hearing").forEach(hearing -> {
+                            EtFortnightlyPressListHelper.formatCaseTime(sitting, hearing);
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java:143]: Same here for reusing method for another list helper

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java (hunk 1 above)
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java]: As discussed, fixed.

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java [hunk 9]
```diff
@@ -217,55 +216,88 @@ private static void formatCaseHtmlTableCrownDailyList(JsonNode hearing) {
...
-                + GeneralHelper.findAndReturnNodeText(individualDetails, "individualSurname")
-                + separator
-                + forNames + " "
-                + middleName).trim();
+            return formatNameWitForNameFirst(title, forNames, middleName, surName);
         }
         return "";
     }
+
+    private static String formatNameWitSurnameFirst(String title, String forName, String middleName, String surname) {
```
[the remark this answers is not in the document]
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java]: fixed.

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java (hunk 5 above)
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java]: fixed.

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java (hunk 6 above)
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java]: fixed.

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java (hunk 7 above)
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java]: fixed.

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java (hunk 8 above)
Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrownFirmListHelper.java:144]: fixed.

File: src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java [hunk 10]
```diff
@@ -217,54 +218,81 @@ private static void formatCaseHtmlTableCrownDailyList(JsonNode hearing) {
...
-        partyDetails = partyDetails
-            + GeneralHelper.stringDelimiter(partyDetails, ", ");
-        builder.insert(0, partyDetails);
+        ((ObjectNode) hearing).put(DEFENDANT, String.join(DELIMITER, defendants));
+        ((ObjectNode) hearing).put(DEFENDANT_REPRESENTATIVE, String.join(DELIMITER, defendantRepresentatives));
+        ((ObjectNode) hearing).put(PROSECUTING_AUTHORITY, String.join(DELIMITER, prosecutingAuthorities));
     }
 
-    private static String createIndividualDetails(JsonNode party) {
+    private static String createIndividualDetails(JsonNode party, boolean surnameFirst) {
```
Comment (KianKwa) [inline src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java:251]: We always have the surname first so don't need the 'surnameFirst' flag anymore.

Comment (junaidiqbalmoj) [reply src/main/java/uk/gov/hmcts/reform/pip/channel/management/services/filegeneration/helpers/listmanipulation/CrimeListHelper.java]: fixed.

**Review by KianKwa — APPROVED** [review]

**Review by ChrisS1512 — APPROVED** [review]

**Pull request merged**